`from_thread.run()` typing issue with TypeVar

### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release


### AnyIO version

3.6.2

### Python version

3.11

### What happened?

Trying to run a program similar to this:

```python
from typing import TypeVar

import anyio
from anyio import to_thread, from_thread

T = TypeVar("T")

async def coro(val: T) -> T:
    return val

def run_from_thread() -> int:
    return from_thread.run(coro, 1)

async def main() -> None:
    val = await to_thread.run_sync(run_from_thread)
    print(val)

if __name__ == "__main__":
    anyio.run(main)
```
mypy and pylance both complain:
```
test.py:12:12: error: Incompatible return value type (got "T", expected "int")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```

### How can we reproduce the bug?

Run the program through the latest mypy (1.1.1)

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -14,7 +14,7 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
   ``TaskGroup.start_soon()``, ``anyio.from_thread.run()``,
   ``anyio.to_thread.run_sync()`` and ``anyio.to_process.run_sync()`` by making use of
   PEP 646 ``TypeVarTuple`` to allow the positional arguments to be validated by static
-  type checkers
+  type checkers (`#560 <https://github.com/agronholm/anyio/issues/560>`_)
 - Fixed adjusting the total number of tokens in a ``CapacityLimiter`` on asyncio failing
   to wake up tasks waiting to acquire the limiter in certain edge cases (fixed with help
   from Egor Blagov)

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -8,6 +8,11 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 - Add support for ``byte``-based paths in ``connect_unix``, ``create_unix_listeners``,
   ``create_unix_datagram_socket``, and ``create_connected_unix_datagram_socket``. (PR by
   Lura Skye)
+- Improved type annotations of numerous methods and functions including ``anyio.run()``,
+  ``TaskGroup.start_soon()``, ``anyio.from_thread.run()``,
+  ``anyio.to_thread.run_sync()`` and ``anyio.to_process.run_sync()`` by making use of
+  PEP 646 ``TypeVarTuple`` to allow the positional arguments to be validated by static
+  type checkers
 - Fixed adjusting the total number of tokens in a ``CapacityLimiter`` on asyncio failing
   to wake up tasks waiting to acquire the limiter in certain edge cases (fixed with help
   from Egor Blagov)
@@ -16,6 +21,7 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 - Fixed cancellation propagating on asyncio from a task group to child tasks if the task
   hosting the task group is in a shielded cancel scope
   (`#642 <https://github.com/agronholm/anyio/issues/642>`_)
+- Fixed the type annotation of ``anyio.Path.samefile()`` to match Typeshed
 
 **4.1.0**
 

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -20,6 +20,7 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
   * ``anyio.from_thread.run()``
   * ``anyio.to_thread.run_sync()``
   * ``anyio.to_process.run_sync()``
+  * ``BlockingPortal.call()``
   * ``BlockingPortal.start_task_soon()``
   * ``BlockingPortal.start_task()``
 

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -10,11 +10,20 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
   Lura Skye)
 - Enabled the ``Event`` and ``CapacityLimiter`` classes to be instantiated outside an
   event loop thread
-- Improved type annotations of numerous methods and functions including ``anyio.run()``,
-  ``TaskGroup.start_soon()``, ``anyio.from_thread.run()``,
-  ``anyio.to_thread.run_sync()`` and ``anyio.to_process.run_sync()`` by making use of
-  PEP 646 ``TypeVarTuple`` to allow the positional arguments to be validated by static
-  type checkers (`#560 <https://github.com/agronholm/anyio/issues/560>`_)
+- Broadly improved/fixed the type annotations. Among other things, many functions and
+  methods that take variadic positional arguments now make use of PEP 646
+  ``TypeVarTuple`` to allow the positional arguments to be validated by static type
+  checkers. These changes affected numerous methods and functions, including:
+
+  * ``anyio.run()``
+  * ``TaskGroup.start_soon()``
+  * ``anyio.from_thread.run()``
+  * ``anyio.to_thread.run_sync()``
+  * ``anyio.to_process.run_sync()``
+  * ``BlockingPortal.start_task_soon()``
+  * ``BlockingPortal.start_task()``
+
+  (`#560 <https://github.com/agronholm/anyio/issues/560>`_)
 - Fixed adjusting the total number of tokens in a ``CapacityLimiter`` on asyncio failing
   to wake up tasks waiting to acquire the limiter in certain edge cases (fixed with help
   from Egor Blagov)

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -30,7 +30,7 @@ dependencies = [
     "exceptiongroup >= 1.0.2; python_version < '3.11'",
     "idna >= 2.8",
     "sniffio >= 1.1",
-    "typing_extensions; python_version < '3.11'",
+    "typing_extensions >= 4.1; python_version < '3.11'",
 ]
 dynamic = ["version"]
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -30,6 +30,7 @@ dependencies = [
     "exceptiongroup >= 1.0.2; python_version < '3.11'",
     "idna >= 2.8",
     "sniffio >= 1.1",
+    "typing_extensions; python_version < '3.11'",
 ]
 dynamic = ["version"]
 

--- a/src/anyio/_backends/_asyncio.py
+++ b/src/anyio/_backends/_asyncio.py
@@ -1964,8 +1964,8 @@ class AsyncIOBackend(AsyncBackend):
     @classmethod
     def run(
         cls,
-        func: Callable[..., Awaitable[T_Retval]],
-        args: tuple,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+        args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         options: dict[str, Any],
     ) -> T_Retval:

--- a/src/anyio/_backends/_asyncio.py
+++ b/src/anyio/_backends/_asyncio.py
@@ -82,8 +82,14 @@ from ..abc import (
 from ..lowlevel import RunVar
 from ..streams.memory import MemoryObjectReceiveStream, MemoryObjectSendStream
 
+if sys.version_info >= (3, 10):
+    from typing import ParamSpec
+else:
+    from typing_extensions import ParamSpec
+
 if sys.version_info >= (3, 11):
     from asyncio import Runner
+    from typing import TypeVarTuple, Unpack
 else:
     import contextvars
     import enum
@@ -91,6 +97,7 @@ else:
     from asyncio import coroutines, events, exceptions, tasks
 
     from exceptiongroup import BaseExceptionGroup
+    from typing_extensions import TypeVarTuple, Unpack
 
     class _State(enum.Enum):
         CREATED = "created"
@@ -271,6 +278,8 @@ else:
 
 T_Retval = TypeVar("T_Retval")
 T_contra = TypeVar("T_contra", contravariant=True)
+PosArgsT = TypeVarTuple("PosArgsT")
+P = ParamSpec("P")
 
 _root_task: RunVar[asyncio.Task | None] = RunVar("_root_task")
 
@@ -682,8 +691,8 @@ class TaskGroup(abc.TaskGroup):
 
     def _spawn(
         self,
-        func: Callable[..., Awaitable[Any]],
-        args: tuple,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[Any]],
+        args: tuple[Unpack[PosArgsT]],
         name: object,
         task_status_future: asyncio.Future | None = None,
     ) -> asyncio.Task:
@@ -752,7 +761,10 @@ class TaskGroup(abc.TaskGroup):
         return task
 
     def start_soon(
-        self, func: Callable[..., Awaitable[Any]], *args: object, name: object = None
+        self,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[Any]],
+        *args: Unpack[PosArgsT],
+        name: object = None,
     ) -> None:
         self._spawn(func, args, name)
 
@@ -875,11 +887,11 @@ class BlockingPortal(abc.BlockingPortal):
 
     def _spawn_task_from_thread(
         self,
-        func: Callable,
-        args: tuple[Any, ...],
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         name: object,
-        future: Future,
+        future: Future[T_Retval],
     ) -> None:
         AsyncIOBackend.run_sync_from_thread(
             partial(self._task_group.start_soon, name=name),
@@ -1883,7 +1895,10 @@ class TestRunner(abc.TestRunner):
                         future.set_result(retval)
 
     async def _call_in_runner_task(
-        self, func: Callable[..., Awaitable[T_Retval]], *args: object, **kwargs: object
+        self,
+        func: Callable[P, Awaitable[T_Retval]],
+        *args: P.args,
+        **kwargs: P.kwargs,
     ) -> T_Retval:
         if not self._runner_task:
             self._send_stream, receive_stream = create_memory_object_stream[
@@ -2062,8 +2077,8 @@ class AsyncIOBackend(AsyncBackend):
     @classmethod
     async def run_sync_in_worker_thread(
         cls,
-        func: Callable[..., T_Retval],
-        args: tuple[Any, ...],
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
         abandon_on_cancel: bool = False,
         limiter: abc.CapacityLimiter | None = None,
     ) -> T_Retval:
@@ -2133,8 +2148,8 @@ class AsyncIOBackend(AsyncBackend):
     @classmethod
     def run_async_from_thread(
         cls,
-        func: Callable[..., Awaitable[T_Retval]],
-        args: tuple[Any, ...],
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+        args: tuple[Unpack[PosArgsT]],
         token: object,
     ) -> T_Retval:
         async def task_wrapper(scope: CancelScope) -> T_Retval:
@@ -2160,7 +2175,10 @@ class AsyncIOBackend(AsyncBackend):
 
     @classmethod
     def run_sync_from_thread(
-        cls, func: Callable[..., T_Retval], args: tuple[Any, ...], token: object
+        cls,
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
+        token: object,
     ) -> T_Retval:
         @wraps(func)
         def wrapper() -> None:

--- a/src/anyio/_backends/_asyncio.py
+++ b/src/anyio/_backends/_asyncio.py
@@ -770,7 +770,7 @@ class TaskGroup(abc.TaskGroup):
 
     async def start(
         self, func: Callable[..., Awaitable[Any]], *args: object, name: object = None
-    ) -> None:
+    ) -> Any:
         future: asyncio.Future = asyncio.Future()
         task = self._spawn(func, args, name, future)
 

--- a/src/anyio/_backends/_asyncio.py
+++ b/src/anyio/_backends/_asyncio.py
@@ -887,7 +887,7 @@ class BlockingPortal(abc.BlockingPortal):
 
     def _spawn_task_from_thread(
         self,
-        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval] | T_Retval],
         args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         name: object,

--- a/src/anyio/_backends/_trio.py
+++ b/src/anyio/_backends/_trio.py
@@ -210,7 +210,7 @@ class BlockingPortal(abc.BlockingPortal):
 
     def _spawn_task_from_thread(
         self,
-        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval] | T_Retval],
         args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         name: object,

--- a/src/anyio/_backends/_trio.py
+++ b/src/anyio/_backends/_trio.py
@@ -62,6 +62,11 @@ from ..abc import IPSockAddrType, UDPPacketType, UNIXDatagramPacketType
 from ..abc._eventloop import AsyncBackend
 from ..streams.memory import MemoryObjectSendStream
 
+if sys.version_info >= (3, 10):
+    from typing import ParamSpec
+else:
+    from typing_extensions import ParamSpec
+
 if sys.version_info >= (3, 11):
     from typing import TypeVarTuple, Unpack
 else:
@@ -72,6 +77,7 @@ T = TypeVar("T")
 T_Retval = TypeVar("T_Retval")
 T_SockAddr = TypeVar("T_SockAddr", str, IPSockAddrType)
 PosArgsT = TypeVarTuple("PosArgsT")
+P = ParamSpec("P")
 
 
 #
@@ -764,9 +770,9 @@ class TestRunner(abc.TestRunner):
 
     def _call_in_runner_task(
         self,
-        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
-        *args: Unpack[PosArgsT],
-        **kwargs: object,
+        func: Callable[P, Awaitable[T_Retval]],
+        *args: P.args,
+        **kwargs: P.kwargs,
     ) -> T_Retval:
         if self._send_stream is None:
             trio.lowlevel.start_guest_run(

--- a/src/anyio/_backends/_trio.py
+++ b/src/anyio/_backends/_trio.py
@@ -192,7 +192,7 @@ class TaskGroup(abc.TaskGroup):
 
     async def start(
         self, func: Callable[..., Awaitable[Any]], *args: object, name: object = None
-    ) -> object:
+    ) -> Any:
         if not self._active:
             raise RuntimeError(
                 "This task group is not active; no new tasks can be started."

--- a/src/anyio/_backends/_trio.py
+++ b/src/anyio/_backends/_trio.py
@@ -62,12 +62,16 @@ from ..abc import IPSockAddrType, UDPPacketType, UNIXDatagramPacketType
 from ..abc._eventloop import AsyncBackend
 from ..streams.memory import MemoryObjectSendStream
 
-if sys.version_info < (3, 11):
+if sys.version_info >= (3, 11):
+    from typing import TypeVarTuple, Unpack
+else:
     from exceptiongroup import BaseExceptionGroup
+    from typing_extensions import TypeVarTuple, Unpack
 
 T = TypeVar("T")
 T_Retval = TypeVar("T_Retval")
 T_SockAddr = TypeVar("T_SockAddr", str, IPSockAddrType)
+PosArgsT = TypeVarTuple("PosArgsT")
 
 
 #
@@ -167,7 +171,12 @@ class TaskGroup(abc.TaskGroup):
         finally:
             self._active = False
 
-    def start_soon(self, func: Callable, *args: object, name: object = None) -> None:
+    def start_soon(
+        self,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[Any]],
+        *args: Unpack[PosArgsT],
+        name: object = None,
+    ) -> None:
         if not self._active:
             raise RuntimeError(
                 "This task group is not active; no new tasks can be started."
@@ -201,11 +210,11 @@ class BlockingPortal(abc.BlockingPortal):
 
     def _spawn_task_from_thread(
         self,
-        func: Callable,
-        args: tuple,
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         name: object,
-        future: Future,
+        future: Future[T_Retval],
     ) -> None:
         trio.from_thread.run_sync(
             partial(self._task_group.start_soon, name=name),
@@ -806,8 +815,8 @@ class TrioBackend(AsyncBackend):
     @classmethod
     def run(
         cls,
-        func: Callable[..., Awaitable[T_Retval]],
-        args: tuple,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+        args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         options: dict[str, Any],
     ) -> T_Retval:
@@ -866,8 +875,8 @@ class TrioBackend(AsyncBackend):
     @classmethod
     async def run_sync_in_worker_thread(
         cls,
-        func: Callable[..., T_Retval],
-        args: tuple[Any, ...],
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
         abandon_on_cancel: bool = False,
         limiter: abc.CapacityLimiter | None = None,
     ) -> T_Retval:
@@ -889,15 +898,18 @@ class TrioBackend(AsyncBackend):
     @classmethod
     def run_async_from_thread(
         cls,
-        func: Callable[..., Awaitable[T_Retval]],
-        args: tuple[Any, ...],
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+        args: tuple[Unpack[PosArgsT]],
         token: object,
     ) -> T_Retval:
         return trio.from_thread.run(func, *args)
 
     @classmethod
     def run_sync_from_thread(
-        cls, func: Callable[..., T_Retval], args: tuple[Any, ...], token: object
+        cls,
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
+        token: object,
     ) -> T_Retval:
         return trio.from_thread.run_sync(func, *args)
 

--- a/src/anyio/_backends/_trio.py
+++ b/src/anyio/_backends/_trio.py
@@ -763,7 +763,10 @@ class TestRunner(abc.TestRunner):
         self._send_stream = None
 
     def _call_in_runner_task(
-        self, func: Callable[..., Awaitable[T_Retval]], *args: object, **kwargs: object
+        self,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+        *args: Unpack[PosArgsT],
+        **kwargs: object,
     ) -> T_Retval:
         if self._send_stream is None:
             trio.lowlevel.start_guest_run(

--- a/src/anyio/_backends/_trio.py
+++ b/src/anyio/_backends/_trio.py
@@ -739,7 +739,7 @@ class TestRunner(abc.TestRunner):
     def __init__(self, **options: Any) -> None:
         from queue import Queue
 
-        self._call_queue: Queue[Callable[..., object]] = Queue()
+        self._call_queue: Queue[Callable[[], object]] = Queue()
         self._send_stream: MemoryObjectSendStream | None = None
         self._options = options
 

--- a/src/anyio/_core/_eventloop.py
+++ b/src/anyio/_core/_eventloop.py
@@ -10,6 +10,11 @@ from typing import TYPE_CHECKING, Any, TypeVar
 
 import sniffio
 
+if sys.version_info >= (3, 11):
+    from typing import TypeVarTuple, Unpack
+else:
+    from typing_extensions import TypeVarTuple, Unpack
+
 if TYPE_CHECKING:
     from ..abc import AsyncBackend
 
@@ -17,12 +22,14 @@ if TYPE_CHECKING:
 BACKENDS = "asyncio", "trio"
 
 T_Retval = TypeVar("T_Retval")
+PosArgsT = TypeVarTuple("PosArgsT")
+
 threadlocals = threading.local()
 
 
 def run(
-    func: Callable[..., Awaitable[T_Retval]],
-    *args: object,
+    func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+    *args: Unpack[PosArgsT],
     backend: str = "asyncio",
     backend_options: dict[str, Any] | None = None,
 ) -> T_Retval:

--- a/src/anyio/_core/_fileio.py
+++ b/src/anyio/_core/_fileio.py
@@ -15,7 +15,6 @@ from typing import (
     AsyncIterator,
     Final,
     Generic,
-    cast,
     overload,
 )
 
@@ -211,7 +210,7 @@ class _PathIterator(AsyncIterator["Path"]):
         if nextval is None:
             raise StopAsyncIteration from None
 
-        return Path(cast("PathLike[str]", nextval))
+        return Path(nextval)
 
 
 class Path:
@@ -518,7 +517,7 @@ class Path:
 
     async def readlink(self) -> Path:
         target = await to_thread.run_sync(os.readlink, self._path)
-        return Path(cast(str, target))
+        return Path(target)
 
     async def rename(self, target: str | pathlib.PurePath | Path) -> Path:
         if isinstance(target, Path):
@@ -545,9 +544,7 @@ class Path:
     async def rmdir(self) -> None:
         await to_thread.run_sync(self._path.rmdir)
 
-    async def samefile(
-        self, other_path: str | bytes | int | pathlib.Path | Path
-    ) -> bool:
+    async def samefile(self, other_path: str | PathLike[str]) -> bool:
         if isinstance(other_path, Path):
             other_path = other_path._path
 

--- a/src/anyio/abc/_eventloop.py
+++ b/src/anyio/abc/_eventloop.py
@@ -176,8 +176,8 @@ class AsyncBackend(metaclass=ABCMeta):
     @abstractmethod
     async def run_sync_in_worker_thread(
         cls,
-        func: Callable[..., T_Retval],
-        args: tuple[Any, ...],
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
         abandon_on_cancel: bool = False,
         limiter: CapacityLimiter | None = None,
     ) -> T_Retval:
@@ -192,8 +192,8 @@ class AsyncBackend(metaclass=ABCMeta):
     @abstractmethod
     def run_async_from_thread(
         cls,
-        func: Callable[..., Awaitable[T_Retval]],
-        args: tuple[Any],
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+        args: tuple[Unpack[PosArgsT]],
         token: object,
     ) -> T_Retval:
         pass
@@ -201,7 +201,10 @@ class AsyncBackend(metaclass=ABCMeta):
     @classmethod
     @abstractmethod
     def run_sync_from_thread(
-        cls, func: Callable[..., T_Retval], args: tuple[Any, ...], token: object
+        cls,
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
+        token: object,
     ) -> T_Retval:
         pass
 

--- a/src/anyio/abc/_eventloop.py
+++ b/src/anyio/abc/_eventloop.py
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
 import math
+import sys
 from abc import ABCMeta, abstractmethod
 from collections.abc import AsyncIterator, Awaitable, Mapping
 from os import PathLike
@@ -16,6 +17,11 @@ from typing import (
     TypeVar,
     overload,
 )
+
+if sys.version_info >= (3, 11):
+    from typing import TypeVarTuple, Unpack
+else:
+    from typing_extensions import TypeVarTuple, Unpack
 
 if TYPE_CHECKING:
     from typing import Literal
@@ -39,6 +45,7 @@ if TYPE_CHECKING:
     from ._testing import TestRunner
 
 T_Retval = TypeVar("T_Retval")
+PosArgsT = TypeVarTuple("PosArgsT")
 
 
 class AsyncBackend(metaclass=ABCMeta):
@@ -46,8 +53,8 @@ class AsyncBackend(metaclass=ABCMeta):
     @abstractmethod
     def run(
         cls,
-        func: Callable[..., Awaitable[T_Retval]],
-        args: tuple[Any, ...],
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+        args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         options: dict[str, Any],
     ) -> T_Retval:

--- a/src/anyio/abc/_tasks.py
+++ b/src/anyio/abc/_tasks.py
@@ -1,15 +1,22 @@
 from __future__ import annotations
 
+import sys
 from abc import ABCMeta, abstractmethod
 from collections.abc import Awaitable, Callable
 from types import TracebackType
 from typing import TYPE_CHECKING, Any, Protocol, TypeVar, overload
+
+if sys.version_info >= (3, 11):
+    from typing import TypeVarTuple, Unpack
+else:
+    from typing_extensions import TypeVarTuple, Unpack
 
 if TYPE_CHECKING:
     from .._core._tasks import CancelScope
 
 T_Retval = TypeVar("T_Retval")
 T_contra = TypeVar("T_contra", contravariant=True)
+PosArgsT = TypeVarTuple("PosArgsT")
 
 
 class TaskStatus(Protocol[T_contra]):
@@ -42,8 +49,8 @@ class TaskGroup(metaclass=ABCMeta):
     @abstractmethod
     def start_soon(
         self,
-        func: Callable[..., Awaitable[Any]],
-        *args: object,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[Any]],
+        *args: Unpack[PosArgsT],
         name: object = None,
     ) -> None:
         """

--- a/src/anyio/from_thread.py
+++ b/src/anyio/from_thread.py
@@ -31,7 +31,7 @@ else:
     from typing_extensions import TypeVarTuple, Unpack
 
 T_Retval = TypeVar("T_Retval")
-T_co = TypeVar("T_co")
+T_co = TypeVar("T_co", covariant=True)
 PosArgsT = TypeVarTuple("PosArgsT")
 
 

--- a/src/anyio/from_thread.py
+++ b/src/anyio/from_thread.py
@@ -80,8 +80,8 @@ def run_sync(
 
 
 class _BlockingAsyncContextManager(Generic[T_co], AbstractContextManager):
-    _enter_future: Future
-    _exit_future: Future
+    _enter_future: Future[T_co]
+    _exit_future: Future[bool | None]
     _exit_event: Event
     _exit_exc_info: tuple[
         type[BaseException] | None, BaseException | None, TracebackType | None
@@ -117,8 +117,7 @@ class _BlockingAsyncContextManager(Generic[T_co], AbstractContextManager):
     def __enter__(self) -> T_co:
         self._enter_future = Future()
         self._exit_future = self._portal.start_task_soon(self.run_async_cm)
-        cm = self._enter_future.result()
-        return cast(T_co, cm)
+        return self._enter_future.result()
 
     def __exit__(
         self,

--- a/src/anyio/from_thread.py
+++ b/src/anyio/from_thread.py
@@ -421,7 +421,7 @@ def start_blocking_portal(
     future: Future[BlockingPortal] = Future()
     with ThreadPoolExecutor(1) as executor:
         run_future = executor.submit(
-            _eventloop.run,
+            _eventloop.run,  # type: ignore[arg-type]
             run_portal,
             backend=backend,
             backend_options=backend_options,

--- a/src/anyio/from_thread.py
+++ b/src/anyio/from_thread.py
@@ -336,10 +336,10 @@ class BlockingPortal:
 
     def start_task(
         self,
-        func: Callable[..., Awaitable[Any]],
+        func: Callable[..., Awaitable[T_Retval]],
         *args: object,
         name: object = None,
-    ) -> tuple[Future[Any], Any]:
+    ) -> tuple[Future[T_Retval], Any]:
         """
         Start a task in the portal's task group and wait until it signals for readiness.
 
@@ -351,13 +351,13 @@ class BlockingPortal:
         :return: a tuple of (future, task_status_value) where the ``task_status_value``
             is the value passed to ``task_status.started()`` from within the target
             function
-        :rtype: tuple[concurrent.futures.Future[Any], Any]
+        :rtype: tuple[concurrent.futures.Future[T_Retval], Any]
 
         .. versionadded:: 3.0
 
         """
 
-        def task_done(future: Future) -> None:
+        def task_done(future: Future[T_Retval]) -> None:
             if not task_status_future.done():
                 if future.cancelled():
                     task_status_future.cancel()

--- a/src/anyio/from_thread.py
+++ b/src/anyio/from_thread.py
@@ -193,12 +193,12 @@ class BlockingPortal:
 
     async def _call_func(
         self,
-        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval] | T_Retval],
         args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         future: Future[T_Retval],
     ) -> None:
-        def callback(f: Future) -> None:
+        def callback(f: Future[T_Retval]) -> None:
             if f.cancelled() and self._event_loop_thread_id not in (
                 None,
                 threading.get_ident(),
@@ -206,15 +206,17 @@ class BlockingPortal:
                 self.call(scope.cancel)
 
         try:
-            retval = func(*args, **kwargs)
-            if isawaitable(retval):
+            retval_or_awaitable = func(*args, **kwargs)
+            if isawaitable(retval_or_awaitable):
                 with CancelScope() as scope:
                     if future.cancelled():
                         scope.cancel()
                     else:
                         future.add_done_callback(callback)
 
-                    retval = await retval
+                    retval = await retval_or_awaitable
+            else:
+                retval = retval_or_awaitable
         except self._cancelled_exc_class:
             future.cancel()
             future.set_running_or_notify_cancel()

--- a/src/anyio/from_thread.py
+++ b/src/anyio/from_thread.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import sys
 import threading
 from collections.abc import Awaitable, Callable, Generator
 from concurrent.futures import FIRST_COMPLETED, Future, ThreadPoolExecutor, wait
@@ -24,11 +25,19 @@ from ._core._tasks import CancelScope, create_task_group
 from .abc import AsyncBackend
 from .abc._tasks import TaskStatus
 
+if sys.version_info >= (3, 11):
+    from typing import TypeVarTuple, Unpack
+else:
+    from typing_extensions import TypeVarTuple, Unpack
+
 T_Retval = TypeVar("T_Retval")
 T_co = TypeVar("T_co")
+PosArgsT = TypeVarTuple("PosArgsT")
 
 
-def run(func: Callable[..., Awaitable[T_Retval]], *args: object) -> T_Retval:
+def run(
+    func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]], *args: Unpack[PosArgsT]
+) -> T_Retval:
     """
     Call a coroutine function from a worker thread.
 
@@ -48,7 +57,9 @@ def run(func: Callable[..., Awaitable[T_Retval]], *args: object) -> T_Retval:
     return async_backend.run_async_from_thread(func, args, token=token)
 
 
-def run_sync(func: Callable[..., T_Retval], *args: object) -> T_Retval:
+def run_sync(
+    func: Callable[[Unpack[PosArgsT]], T_Retval], *args: Unpack[PosArgsT]
+) -> T_Retval:
     """
     Call a function in the event loop thread from a worker thread.
 
@@ -182,7 +193,11 @@ class BlockingPortal:
             self._task_group.cancel_scope.cancel()
 
     async def _call_func(
-        self, func: Callable, args: tuple, kwargs: dict[str, Any], future: Future
+        self,
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
+        kwargs: dict[str, Any],
+        future: Future[T_Retval],
     ) -> None:
         def callback(f: Future) -> None:
             if f.cancelled() and self._event_loop_thread_id not in (
@@ -219,11 +234,11 @@ class BlockingPortal:
 
     def _spawn_task_from_thread(
         self,
-        func: Callable,
-        args: tuple[Any, ...],
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         name: object,
-        future: Future,
+        future: Future[T_Retval],
     ) -> None:
         """
         Spawn a new task using the given callable.
@@ -241,17 +256,23 @@ class BlockingPortal:
         raise NotImplementedError
 
     @overload
-    def call(self, func: Callable[..., Awaitable[T_Retval]], *args: object) -> T_Retval:
+    def call(
+        self,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+        *args: Unpack[PosArgsT],
+    ) -> T_Retval:
         ...
 
     @overload
-    def call(self, func: Callable[..., T_Retval], *args: object) -> T_Retval:
+    def call(
+        self, func: Callable[[Unpack[PosArgsT]], T_Retval], *args: Unpack[PosArgsT]
+    ) -> T_Retval:
         ...
 
     def call(
         self,
-        func: Callable[..., Awaitable[T_Retval] | T_Retval],
-        *args: object,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval] | T_Retval],
+        *args: Unpack[PosArgsT],
     ) -> T_Retval:
         """
         Call the given function in the event loop thread.
@@ -268,22 +289,25 @@ class BlockingPortal:
     @overload
     def start_task_soon(
         self,
-        func: Callable[..., Awaitable[T_Retval]],
-        *args: object,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval]],
+        *args: Unpack[PosArgsT],
         name: object = None,
     ) -> Future[T_Retval]:
         ...
 
     @overload
     def start_task_soon(
-        self, func: Callable[..., T_Retval], *args: object, name: object = None
+        self,
+        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        *args: Unpack[PosArgsT],
+        name: object = None,
     ) -> Future[T_Retval]:
         ...
 
     def start_task_soon(
         self,
-        func: Callable[..., Awaitable[T_Retval] | T_Retval],
-        *args: object,
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval] | T_Retval],
+        *args: Unpack[PosArgsT],
         name: object = None,
     ) -> Future[T_Retval]:
         """

--- a/src/anyio/from_thread.py
+++ b/src/anyio/from_thread.py
@@ -233,7 +233,7 @@ class BlockingPortal:
 
     def _spawn_task_from_thread(
         self,
-        func: Callable[[Unpack[PosArgsT]], T_Retval],
+        func: Callable[[Unpack[PosArgsT]], Awaitable[T_Retval] | T_Retval],
         args: tuple[Unpack[PosArgsT]],
         kwargs: dict[str, Any],
         name: object,
@@ -328,7 +328,7 @@ class BlockingPortal:
 
         """
         self._check_running()
-        f: Future = Future()
+        f: Future[T_Retval] = Future()
         self._spawn_task_from_thread(func, args, {}, name, f)
         return f
 

--- a/src/anyio/streams/tls.py
+++ b/src/anyio/streams/tls.py
@@ -3,6 +3,7 @@ from __future__ import annotations
 import logging
 import re
 import ssl
+import sys
 from collections.abc import Callable, Mapping
 from dataclasses import dataclass
 from functools import wraps
@@ -17,7 +18,13 @@ from .. import (
 from .._core._typedattr import TypedAttributeSet, typed_attribute
 from ..abc import AnyByteStream, ByteStream, Listener, TaskGroup
 
+if sys.version_info >= (3, 11):
+    from typing import TypeVarTuple, Unpack
+else:
+    from typing_extensions import TypeVarTuple, Unpack
+
 T_Retval = TypeVar("T_Retval")
+PosArgsT = TypeVarTuple("PosArgsT")
 _PCTRTT = Tuple[Tuple[str, str], ...]
 _PCTRTTT = Tuple[_PCTRTT, ...]
 
@@ -126,7 +133,7 @@ class TLSStream(ByteStream):
         return wrapper
 
     async def _call_sslobject_method(
-        self, func: Callable[..., T_Retval], *args: object
+        self, func: Callable[[Unpack[PosArgsT]], T_Retval], *args: Unpack[PosArgsT]
     ) -> T_Retval:
         while True:
             try:

--- a/src/anyio/to_process.py
+++ b/src/anyio/to_process.py
@@ -18,9 +18,16 @@ from .abc import ByteReceiveStream, ByteSendStream, Process
 from .lowlevel import RunVar, checkpoint_if_cancelled
 from .streams.buffered import BufferedByteReceiveStream
 
+if sys.version_info >= (3, 11):
+    from typing import TypeVarTuple, Unpack
+else:
+    from typing_extensions import TypeVarTuple, Unpack
+
 WORKER_MAX_IDLE_TIME = 300  # 5 minutes
 
 T_Retval = TypeVar("T_Retval")
+PosArgsT = TypeVarTuple("PosArgsT")
+
 _process_pool_workers: RunVar[set[Process]] = RunVar("_process_pool_workers")
 _process_pool_idle_workers: RunVar[deque[tuple[Process, float]]] = RunVar(
     "_process_pool_idle_workers"
@@ -29,8 +36,8 @@ _default_process_limiter: RunVar[CapacityLimiter] = RunVar("_default_process_lim
 
 
 async def run_sync(
-    func: Callable[..., T_Retval],
-    *args: object,
+    func: Callable[[Unpack[PosArgsT]], T_Retval],
+    *args: Unpack[PosArgsT],
     cancellable: bool = False,
     limiter: CapacityLimiter | None = None,
 ) -> T_Retval:

--- a/src/anyio/to_thread.py
+++ b/src/anyio/to_thread.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import sys
 from collections.abc import Callable
 from typing import TypeVar
 from warnings import warn
@@ -7,12 +8,18 @@ from warnings import warn
 from ._core._eventloop import get_async_backend
 from .abc import CapacityLimiter
 
+if sys.version_info >= (3, 11):
+    from typing import TypeVarTuple, Unpack
+else:
+    from typing_extensions import TypeVarTuple, Unpack
+
 T_Retval = TypeVar("T_Retval")
+PosArgsT = TypeVarTuple("PosArgsT")
 
 
 async def run_sync(
-    func: Callable[..., T_Retval],
-    *args: object,
+    func: Callable[[Unpack[PosArgsT]], T_Retval],
+    *args: Unpack[PosArgsT],
     abandon_on_cancel: bool = False,
     cancellable: bool | None = None,
     limiter: CapacityLimiter | None = None,

--- a/tests/test_from_thread.py
+++ b/tests/test_from_thread.py
@@ -206,8 +206,8 @@ class TestRunAsyncFromThread:
         exc.match("unsupported operand type")
 
     async def test_run_anyio_async_func_from_thread(self) -> None:
-        def worker(*args: int) -> Literal[True]:
-            from_thread.run(sleep, *args)
+        def worker(delay: float) -> Literal[True]:
+            from_thread.run(sleep, delay)
             return True
 
         assert await to_thread.run_sync(worker, 0)

--- a/tests/test_from_thread.py
+++ b/tests/test_from_thread.py
@@ -507,29 +507,29 @@ class TestBlockingPortal:
     def test_start_no_value(
         self, anyio_backend_name: str, anyio_backend_options: dict[str, Any]
     ) -> None:
-        def taskfunc(*, task_status: TaskStatus) -> None:
+        async def taskfunc(*, task_status: TaskStatus) -> None:
             task_status.started()
 
         with start_blocking_portal(anyio_backend_name, anyio_backend_options) as portal:
-            future, value = portal.start_task(taskfunc)  # type: ignore[arg-type]
+            future, value = portal.start_task(taskfunc)
             assert value is None
             assert future.result() is None
 
     def test_start_with_value(
         self, anyio_backend_name: str, anyio_backend_options: dict[str, Any]
     ) -> None:
-        def taskfunc(*, task_status: TaskStatus) -> None:
+        async def taskfunc(*, task_status: TaskStatus) -> None:
             task_status.started("foo")
 
         with start_blocking_portal(anyio_backend_name, anyio_backend_options) as portal:
-            future, value = portal.start_task(taskfunc)  # type: ignore[arg-type]
+            future, value = portal.start_task(taskfunc)
             assert value == "foo"
             assert future.result() is None
 
     def test_start_crash_before_started_call(
         self, anyio_backend_name: str, anyio_backend_options: dict[str, Any]
     ) -> None:
-        def taskfunc(*, task_status: object) -> NoReturn:
+        async def taskfunc(*, task_status: object) -> NoReturn:
             raise Exception("foo")
 
         with start_blocking_portal(anyio_backend_name, anyio_backend_options) as portal:
@@ -539,7 +539,7 @@ class TestBlockingPortal:
     def test_start_crash_after_started_call(
         self, anyio_backend_name: str, anyio_backend_options: dict[str, Any]
     ) -> None:
-        def taskfunc(*, task_status: TaskStatus) -> NoReturn:
+        async def taskfunc(*, task_status: TaskStatus) -> NoReturn:
             task_status.started(2)
             raise Exception("foo")
 
@@ -552,24 +552,21 @@ class TestBlockingPortal:
     def test_start_no_started_call(
         self, anyio_backend_name: str, anyio_backend_options: dict[str, Any]
     ) -> None:
-        def taskfunc(*, task_status: TaskStatus) -> None:
+        async def taskfunc(*, task_status: TaskStatus) -> None:
             pass
 
         with start_blocking_portal(anyio_backend_name, anyio_backend_options) as portal:
             with pytest.raises(RuntimeError, match="Task exited"):
-                portal.start_task(taskfunc)  # type: ignore[arg-type]
+                portal.start_task(taskfunc)
 
     def test_start_with_name(
         self, anyio_backend_name: str, anyio_backend_options: dict[str, Any]
     ) -> None:
-        def taskfunc(*, task_status: TaskStatus) -> None:
+        async def taskfunc(*, task_status: TaskStatus) -> None:
             task_status.started(get_current_task().name)
 
         with start_blocking_portal(anyio_backend_name, anyio_backend_options) as portal:
-            future, start_value = portal.start_task(
-                taskfunc,  # type: ignore[arg-type]
-                name="testname",
-            )
+            future, start_value = portal.start_task(taskfunc, name="testname")
             assert start_value == "testname"
 
     def test_contextvar_propagation_sync(

